kubernetes_env/benchmark/loop_filter/filter.cc

```diff
@@ -0,0 +1,151 @@
+  }
+  return TrafficDirection::Unspecified;
+}
+
+std::string trafficDirectionToString(TrafficDirection dir) {
+  if (dir == TrafficDirection::Unspecified) {
+    return "unspecified";
+  } else if (dir == TrafficDirection::Inbound) {
+    return "inbound";
+  } else {
+    return "outbound";
+  }
+}
+
+// udf_type: Scalar
+// id: count
+// return_type: int
+
+class BidiRootContext : public RootContext {
+public:
+  explicit BidiRootContext(uint32_t id, std::string_view root_id)
+      : RootContext(id, root_id) {
+    std::string workload_name;
+    if (getValue({"node", "metadata", "WORKLOAD_NAME"}, &workload_name)) {
+      workload_name_ = workload_name;
+      LOG_WARN("Initialized workload_name: " + workload_name_);
+    } else {
+      LOG_WARN("Failed to set workload name");
+    }
+  }
+  bool onConfigure(size_t /* configuration_size */) override;
+
+  std::string_view getWorkloadName() { return workload_name_; }
+
+private:
+  std::string workload_name_;
+};
+
+class BidiContext : public Context {
+public:
+  explicit BidiContext(uint32_t id, RootContext *root)
+      : Context(id, root),
+        root_(static_cast<BidiRootContext *>(static_cast<void *>(root))) {
+    direction_ = getTrafficDirection();
+  }
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override;
+  FilterHeadersStatus onRequestHeadersInbound();
+  FilterHeadersStatus onRequestHeadersOutbound();
+  FilterHeadersStatus onResponseHeaders(uint32_t headers,
+                                        bool end_of_stream) override;
+  FilterHeadersStatus onResponseHeadersInbound();
+  FilterHeadersStatus onResponseHeadersOutbound();
+
+private:
+  BidiRootContext *root_;
+  TrafficDirection direction_;
+};
+
+static RegisterContextFactory
+    register_BidiContext(CONTEXT_FACTORY(BidiContext),
+                         ROOT_FACTORY(BidiRootContext), "bidi_root_id");
+
+bool BidiRootContext::onConfigure(size_t) { return true; }
+
+FilterHeadersStatus BidiContext::onRequestHeaders(uint32_t, bool) {
+
+  if (direction_ == TrafficDirection::Inbound) {
+    return onRequestHeadersInbound();
+  } else if (direction_ == TrafficDirection::Outbound) {
+    return onRequestHeadersOutbound();
+  }
+}
+
+FilterHeadersStatus BidiContext::onRequestHeadersInbound() {
+  LOG_WARN("Traversed.");
+  int counter = 0;
+  int max = 1000;
+  int start_s = clock(); 
+  for(int i = 0; i < max; i++) {
+    counter += 1;
+    std::cout << counter << "\n";
+  }
+  int stop_s = clock();
+  double timestamp = (stop_s - start_s) / double(CLOCKS_PER_SEC);
+  std::string time = "Time" + std::to_string(timestamp);
+  LOG_WARN(time);
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus BidiContext::onRequestHeadersOutbound() {
+  LOG_WARN("Traversed.");
+  int counter = 1000;
+  int start_s = clock();
+  for(int i = 0; i >= 0; i++) {
+    counter -= 1;
+    std::cout << counter << "\n";
```

Comment:
You may want to use the built-in logging facilities instead of cout. I am actually surprised this prints. Or this line intentional to slow down the filter?